### PR TITLE
SHARD-2389 - ShardFunctions unit tests

### PR DESCRIPTION
**Pull request by S0naliThakur**
### **PR Type**
Tests

___

### **Description**
- Added comprehensive unit tests for `ShardFunctions`.

- Tested various scenarios including edge cases and error handling.

- Implemented mock data setup for consistent test environments.

- Verified functionality of partition and node-related calculations.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ShardFunctions.test.ts</strong><dd><code>Comprehensive unit tests for ShardFunctions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

test/unit/src/ShardFunctions.test.ts

<li>Added 2777 lines of unit tests for <code>ShardFunctions</code>.<br> <li> Tested functions like <code>calculateShardGlobals</code>, <code>leadZeros8</code>, and <br><code>computeCoverageChanges</code>.<br> <li> Included tests for edge cases and error handling.<br> <li> Utilized mock data for nodes and partitions.

</details>

  </td>
  <td><a href="https://github.com/shardeum/archiver/pull/187/files#diff-e5dcbac0589c6671ee473b29041b3e8c61d6c5b342ced8e9652caa6de6766bd9">+2777/-0</a></td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: test/unit/src/ShardFunctions.test.ts [hunk 1]
```diff
@@ -0,0 +1,2777 @@
...
+      expect(result.numActiveNodes).toBe(10)
+      expect(result.nodesPerConsenusGroup).toBe(3)
+      expect(result.numPartitions).toBe(10)
+      expect(result.consensusRadius).toBe(1)
+      expect(result.nodesPerEdge).toBe(1)
+      expect(result.numVisiblePartitions).toBe(5)
+      expect(typeof result.nodeLookRange).toBe('number')
+    })
+
+    it('should handle odd nodesPerConsensusGroup value', () => {
```
Comment (aniketdivekar) [inline test/unit/src/ShardFunctions.test.ts:135]: we should also add a test w/ even value and see if we get correct output - after all, its a valid test case

File: test/unit/src/ShardFunctions.test.ts [hunk 2]
```diff
@@ -0,0 +1,2777 @@
...
+
+      const result = ShardFunctions.calculateShardGlobals(numNodes, nodesPerConsensusGroup, nodesPerEdge)
+
+      expect(result.nodesPerEdge).toBe(2) // Should match consensusRadius
+    })
+  })
+
+  describe('#leadZeros8', () => {
+    it('should pad a string with leading zeros to make it 8 characters', () => {
+      expect(ShardFunctions.leadZeros8('1')).toBe('00000001')
```
Comment (aniketdivekar) [inline test/unit/src/ShardFunctions.test.ts:169]: any case on whats the max length string `leadZeros8` method can accept? like what if I pass string of length 1000? 
is that case handled in the code? if not, we should add a test and comment it. Also create a ticket under SHARD-1967 and put it in triage

File: test/unit/src/ShardFunctions.test.ts [hunk 3]
```diff
@@ -0,0 +1,2777 @@
...
+    it('should calculate shard values for a given address', () => {
+      const address = '12345678deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef'
+
+      const result = ShardFunctions.calculateShardValues(mockShardGlobals, address)
+
+      expect(result).toBeDefined()
+      expect(result.address).toBe(address)
+      expect(result.addressPrefix).toBe(parseInt('12345678', 16))
+      expect(result.addressPrefixHex).toBe('12345678')
+      expect(typeof result.homePartition).toBe('number')
```
Comment (aniketdivekar) [inline test/unit/src/ShardFunctions.test.ts:189]: possible to check against its value? is it deterministic ?

File: test/unit/src/ShardFunctions.test.ts [hunk 4]
```diff
@@ -0,0 +1,2777 @@
...
+      expect(result.homeRange).toBeDefined()
+      expect(result.coveredBy).toEqual({})
+      expect(result.storedBy).toEqual({})
+    })
+  })
+
+  describe('#calculateStoredPartitions2', () => {
+    it('should calculate stored partitions for a given home partition', () => {
```
Comment (aniketdivekar) [inline test/unit/src/ShardFunctions.test.ts:197]: can we have some negative tests here?

File: test/unit/src/ShardFunctions.test.ts [hunk 5]
```diff
@@ -0,0 +1,2777 @@
...
+      const result = ShardFunctions.calculateStoredPartitions2(mockShardGlobals, homePartition)
+
+      expect(result).toBeDefined()
+      expect(result.homeRange).toBeDefined()
+      expect(typeof result.rangeIsSplit).toBe('boolean')
+    })
+  })
+
+  describe('#calculateConsensusPartitions', () => {
+    it('should calculate consensus partitions for a given home partition', () => {
```
Comment (aniketdivekar) [inline test/unit/src/ShardFunctions.test.ts:209]: > can we have some negative tests here?

same here

File: test/unit/src/ShardFunctions.test.ts [hunk 6]
```diff
@@ -0,0 +1,2777 @@
...
+  })
+
+  describe('#addressToPartition', () => {
+    it('should convert address to partition', () => {
+      const address = '12345678deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef'
+
+      const result = ShardFunctions.addressToPartition(mockShardGlobals, address)
+
+      expect(result).toBeDefined()
+      expect(typeof result.homePartition).toBe('number')
```
Comment (aniketdivekar) [inline test/unit/src/ShardFunctions.test.ts:640]: can we also validate the value in addition to type? is the output deterministic ?
